File: expense-report/python/src/expense_report/report.py
from __future__ import annotations

from functools import reduce

from .expense_item import ExpenseItem
from .money import Money
from .report_status import ReportStatus
from . import spending_policy
from .exceptions import (
    EmptyReportError,
    FinalizedReportError,
    InvalidAmountError,
    InvalidStatusTransitionError,
    ReportExceedsMaximumError,
    ExpenseNotFoundError,
)


class Report:
    def __init__(self, employee_name: str) -> None:
        self._employee_name = employee_name
        self._expenses: list[ExpenseItem] = []
        self._status = ReportStatus.DRAFT
        self._rejection_reason: str | None = None

    @property
    def employee_name(self) -> str:
        return self._employee_name

    @property
    def status(self) -> ReportStatus:
        return self._status

    @property
    def expenses(self) -> list[ExpenseItem]:
        return list(self._expenses)

    @property
    def rejection_reason(self) -> str | None:
        return self._rejection_reason

    @property
    def total(self) -> Money:
        return reduce(lambda s, e: s + e.amount, self._expenses, Money.zero())

    @property
    def requires_approval(self) -> bool:
        return any(e.is_over_limit for e in self._expenses) or self.total > spending_policy.APPROVAL_THRESHOLD

    @property
    def approval_reason(self) -> str:
        if any(e.is_over_limit for e in self._expenses):
            return "Yes (over-limit items)"
        if self.total > spending_policy.APPROVAL_THRESHOLD:
            return "Yes (total exceeds $2,500)"
        return "No"

    def add_expense(self, item: ExpenseItem) -> None:
        if self._status in (ReportStatus.APPROVED, ReportStatus.REJECTED):
            raise FinalizedReportError()
        if not item.amount.is_positive:
            raise InvalidAmountError()
        self._expenses.append(item)

    def remove_expense(self, item: ExpenseItem) -> None:
        try:
            self._expenses.remove(item)
        except ValueError:
            raise ExpenseNotFoundError() from None

    def submit(self) -> None:
        if not self._expenses:
            raise EmptyReportError()
        if self.total > spending_policy.REPORT_MAXIMUM:
            raise ReportExceedsMaximumError()
        self._status = ReportStatus.PENDING

    def approve(self) -> None:
        if self._status != ReportStatus.PENDING:
            raise InvalidStatusTransitionError("Only pending reports can be approved")
        self._status = ReportStatus.APPROVED

    def reject(self, reason: str) -> None:
        if self._status != ReportStatus.PENDING:
            raise InvalidStatusTransitionError("Only pending reports can be rejected")
        self._status = ReportStatus.REJECTED
        self._rejection_reason = reason

    def reopen(self) -> None:
        if self._status != ReportStatus.REJECTED:
            raise InvalidStatusTransitionError("Only rejected reports can be reopened")
        self._status = ReportStatus.DRAFT
        self._rejection_reason = None

    def print_summary(self) -> str:
        lines: list[str] = []
        lines.append(f"Expense Report: {self._employee_name}")
        lines.append(f"Status: {self._status.value}")
        lines.append("")
        for e in self._expenses:
            flag = " [OVER LIMIT]" if e.is_over_limit else ""
            amount = e.amount.format()
            lines.append(f"  {e.category.value}: {e.description}  ${amount}{flag}")
        lines.append("")
        lines.append(f"Total: ${self.total.format()}")
        lines.append(f"Requires Approval: {self.approval_reason}")
        return "\n".join(lines)

File: expense-report/typescript/src/Report.ts
import type { ExpenseItem } from './ExpenseItem.js';
import { Money } from './Money.js';
import { ReportStatus } from './ReportStatus.js';
import { SpendingPolicy } from './SpendingPolicy.js';
import {
  EmptyReportError,
  FinalizedReportError,
  InvalidAmountError,
  InvalidStatusTransitionError,
  ReportExceedsMaximumError,
  ExpenseNotFoundError,
} from './Exceptions.js';

export class Report {
  private _expenses: ExpenseItem[] = [];
  private _status: ReportStatus = ReportStatus.Draft;
  private _rejectionReason: string | null = null;

  constructor(public readonly employeeName: string) {}

  get status(): ReportStatus { return this._status; }
  get expenses(): readonly ExpenseItem[] { return this._expenses; }
  get rejectionReason(): string | null { return this._rejectionReason; }

  get total(): Money {
    return this._expenses.reduce((sum, e) => sum.plus(e.amount), Money.zero);
  }

  get requiresApproval(): boolean {
    return this._expenses.some(e => e.isOverLimit) || this.total.greaterThan(SpendingPolicy.approvalThreshold);
  }

  get approvalReason(): string {
    if (this._expenses.some(e => e.isOverLimit)) return 'Yes (over-limit items)';
    if (this.total.greaterThan(SpendingPolicy.approvalThreshold)) return 'Yes (total exceeds $2,500)';
    return 'No';
  }

  addExpense(item: ExpenseItem): void {
    if (this._status === ReportStatus.Approved || this._status === ReportStatus.Rejected)
      throw new FinalizedReportError();
    if (!item.amount.isPositive)
      throw new InvalidAmountError();
    this._expenses.push(item);
  }

  removeExpense(item: ExpenseItem): void {
    const index = this._expenses.indexOf(item);
    if (index === -1) throw new ExpenseNotFoundError();
    this._expenses.splice(index, 1);
  }

  submit(): void {
    if (this._expenses.length === 0)
      throw new EmptyReportError();
    if (this.total.greaterThan(SpendingPolicy.reportMaximum))
      throw new ReportExceedsMaximumError();
    this._status = ReportStatus.Pending;
  }

  approve(): void {
    if (this._status !== ReportStatus.Pending)
      throw new InvalidStatusTransitionError('Only pending reports can be approved');
    this._status = ReportStatus.Approved;
  }

  reject(reason: string): void {
    if (this._status !== ReportStatus.Pending)
      throw new InvalidStatusTransitionError('Only pending reports can be rejected');
    this._status = ReportStatus.Rejected;
    this._rejectionReason = reason;
  }

  reopen(): void {
    if (this._status !== ReportStatus.Rejected)
      throw new InvalidStatusTransitionError('Only rejected reports can be reopened');
    this._status = ReportStatus.Draft;
    this._rejectionReason = null;
  }

  printSummary(): string {
    const lines: string[] = [];
    lines.push(`Expense Report: ${this.employeeName}`);
    lines.push(`Status: ${this._status}`);
    lines.push('');
    for (const e of this._expenses) {
      const flag = e.isOverLimit ? ' [OVER LIMIT]' : '';
      const amount = e.amount.format();
      lines.push(`  ${e.category}: ${e.description}  $${amount}${flag}`);
    }
    lines.push('');
    lines.push(`Total: $${this.total.format()}`);
    lines.push(`Requires Approval: ${this.approvalReason}`);
    return lines.join('\n');
  }
}
